datasources/src/main/java/org/opensearch/sql/datasources/storage/OpenSearchDataSourceMetadataStorage.java

```diff
@@ -86,6 +90,56 @@ public List<DataSourceMetadata> getDataSourceMetadata() {
     return searchInDataSourcesIndex(QueryBuilders.matchAllQuery());
   }
 
+  /**
+   * Counts the number of data sources per connector type in the specified OpenSearch index. This
+   * method queries the OpenSearch index defined by {@code DATASOURCE_INDEX_NAME} and aggregates the
+   * count of documents (data sources) based on the connector type, which is expected to be
+   * specified in the document field {@code connector}.
+   *
+   * <p>Note: This method uses the client's thread pool to execute the search request and
+   * temporarily stashes the current thread context to avoid context leakage.
+   *
+   * @return A {@link Map} where keys are connector types (as {@link String}) and values are counts
+   *     (as {@link Long}) of data sources of each type. Returns an empty map if the index does not
+   *     exist or if no data sources are found.
+   * @throws RuntimeException if the search operation fails or returns a non-200 status code,
+   *     indicating an error with the Elasticsearch query or connection.
+   */
+  @Override
+  public Map<String, Long> countDataSourcesPerConnector() {
+    if (!this.clusterService.state().routingTable().hasIndex(DATASOURCE_INDEX_NAME)) {
+      createDataSourcesIndex();
+      return Collections.emptyMap();
+    }
+
+    SearchRequest searchRequest = new SearchRequest(DATASOURCE_INDEX_NAME);
```

Comment:
is there any security concern to expose datasource info through ppl/stats api? @vamsi-amazon .
btw, does `GET datasource` api achieve same capability?

Reply:
This particular addition is not a security concern in my opinion. It is just the count of datasources.

Overall stats API is not guarded by any permissions. This should be a concern as we are exposing a lot of information.

@noCharger  Can you raise an issue for this? PPL stats api is open right now.

Reply:
> @noCharger can you refer other plugin stats API. are they open or guarded by permissions?

@vamsi-amazon  Ref OS core - https://opensearch.org/docs/latest/api-reference/nodes-apis/nodes-stats/#required-permissions
But I can still get plugin stats access with `datasource_read_role | cluster:admin/opensearch/ql/datasources/read `

Reply:
> This particular addition is not a security concern in my opinion. It is just the count of datasources.
> 
> Overall stats API is not guarded by any permissions. This should be a concern as we are exposing a lot of information.
> 
> @noCharger Can you raise an issue for this? PPL stats api is open right now.

https://github.com/opensearch-project/sql/issues/2498